File: src/sync.rs
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::mpsc;

use crate::lyrics::{ensure_leading_line, Line, LyricsProvider};
use crate::player::Player;
use crate::renderer::{DisplayLine, Update};

pub struct SyncConfig {
    pub player_poll_interval: Duration,
    pub ui_timer_interval: Duration,
    pub merge_queue: bool,
}

impl Default for SyncConfig {
    fn default() -> Self {
        Self {
            player_poll_interval: Duration::from_millis(2000),
            ui_timer_interval: Duration::from_millis(200),
            merge_queue: false,
        }
    }
}

pub fn start_sync(
    player: Arc<dyn Player>,
    provider: Box<dyn LyricsProvider>,
    config: SyncConfig,
) -> mpsc::Receiver<Update> {
    let (tx, rx) = mpsc::channel(16);
    tokio::spawn(sync_loop(player, provider, config, tx));
    rx
}

struct SyncState {
    current_track_id: String,
    current_lyrics: Vec<Line>,
    display_lines: Vec<DisplayLine>,
    current_track_offset: usize,
    current_duration: u64,
    next_lyrics: Vec<Line>,
    next_artist: String,
    next_track_name: String,
    next_track_id: String,
    next_track_start: Option<usize>,
    index: usize,
    last_state_position: u64,
    last_state_time: Instant,
    is_playing: bool,
    error: Option<String>,
    queue_rechecked: bool,
}

impl Default for SyncState {
    fn default() -> Self {
        Self {
            current_track_id: String::new(),
            current_lyrics: Vec::new(),
            display_lines: Vec::new(),
            current_track_offset: 0,
            current_duration: 0,
            next_lyrics: Vec::new(),
            next_artist: String::new(),
            next_track_name: String::new(),
            next_track_id: String::new(),
            next_track_start: None,
            index: 0,
            last_state_position: 0,
            last_state_time: Instant::now(),
            is_playing: false,
            error: None,
            queue_rechecked: false,
        }
    }
}

impl SyncState {
    fn interpolated_position(&self) -> u64 {
        if self.is_playing {
            self.last_state_position + self.last_state_time.elapsed().as_millis() as u64
        } else {
            self.last_state_position
        }
    }

    fn lyrics_to_display(lyrics: &[Line]) -> Vec<DisplayLine> {
        lyrics.iter().cloned().map(DisplayLine::Lyric).collect()
    }

    fn clear(&mut self) {
        self.current_track_id.clear();
        self.current_lyrics.clear();
        self.display_lines.clear();
        self.next_lyrics.clear();
        self.next_track_id.clear();
        self.next_track_start = None;
        self.current_track_offset = 0;
        self.current_duration = 0;
        self.index = 0;
        self.error = None;
    }

    fn to_update(&self) -> Update {
        Update {
            lines: self.display_lines.clone(),
            index: self.index,
            error: self.error.clone(),
        }
    }

    async fn fetch_next_track(
        &mut self,
        player: &dyn Player,
        provider: &dyn LyricsProvider,
        merge_queue: bool,
    ) {
        self.next_lyrics.clear();
        self.next_track_id.clear();
        self.next_track_start = None;
        if let Ok(queue) = player.queue().await {
            if let Some(next) = queue.into_iter().next() {
                self.next_track_id = next.track_id;
                self.next_artist = next.artist;
                self.next_track_name = next.track;
                if let Ok(nl) = provider.fetch(&self.next_artist, &self.next_track_name).await {
                    self.next_lyrics = nl;
                    ensure_leading_line(&mut self.next_lyrics);
                    if merge_queue {
                        self.next_track_start = append_next_track(
                            &mut self.display_lines,
                            &self.next_artist,
                            &self.next_track_name,
                            &self.next_lyrics,
                        );
                    }
                }
            }
        }
    }

    async fn transition_to_new_track(
        &mut self,
        track_id: &str,
        artist: &str,
        track: &str,
        provider: &dyn LyricsProvider,
        player: &dyn Player,
        smooth: bool,
        merge_queue: bool,
    ) {
        if smooth {
            self.current_track_offset = self.next_track_start.unwrap();
            self.current_lyrics = std::mem::take(&mut self.next_lyrics);
        } else if track_id == self.next_track_id && !self.next_lyrics.is_empty() {
            self.current_track_offset = 0;
            self.current_lyrics = std::mem::take(&mut self.next_lyrics);
            self.display_lines = Self::lyrics_to_display(&self.current_lyrics);
        } else {
            self.current_track_offset = 0;
            match provider.fetch(artist, track).await {
                Ok(l) => {
                    self.current_lyrics = l;
                    ensure_leading_line(&mut self.current_lyrics);
                }
                Err(e) => {
                    self.current_lyrics.clear();
                    self.display_lines.clear();
                    self.next_lyrics.clear();
                    self.next_track_id.clear();
                    self.next_track_start = None;
                    self.error = Some(e.to_string());
                    self.current_track_id = track_id.to_string();
                    self.index = 0;
                    return;
                }
            }
            self.display_lines = Self::lyrics_to_display(&self.current_lyrics);
        }

        self.current_track_id = track_id.to_string();
        self.index = self.current_track_offset;
        self.error = None;

        self.fetch_next_track(player, provider, merge_queue).await;
    }
}

async fn sync_loop(
    player: Arc<dyn Player>,
    provider: Box<dyn LyricsProvider>,
    config: SyncConfig,
    tx: mpsc::Sender<Update>,
) {
    let (player_tx, mut player_rx) = mpsc::channel(1);

    let player_poll = player.clone();
    let poll_interval = config.player_poll_interval;
    tokio::spawn(async move {
        loop {
            let result = player_poll.state().await;
            if player_tx.send(result).await.is_err() {
                break;
            }
            tokio::time::sleep(poll_interval).await;
        }
    });

    let mut state = SyncState::default();
    let mut sleep_duration = config.ui_timer_interval;

    loop {
        tokio::select! {
            Some(result) = player_rx.recv() => {
                match result {
                    Ok(Some(player_state)) => {
                        state.is_playing = player_state.is_playing;
                        state.current_duration = player_state.duration_ms;

                        if player_state.track_id != state.current_track_id {
                            let smooth = player_state.track_id == state.next_track_id
                                && state.next_track_start.is_some();

                            state.transition_to_new_track(
                                &player_state.track_id,
                                &player_state.artist,
                                &player_state.track,
                                &*provider,
                                &*player,
                                smooth && config.merge_queue,
                                config.merge_queue,
                            ).await;
                            state.queue_rechecked = false;
                        }

                        state.last_state_position = player_state.position_ms;
                        state.last_state_time = Instant::now();
                    }
                    Ok(None) => {
                        state.is_playing = false;
                        if !state.current_track_id.is_empty() {
                            state.clear();
                        }
                    }
                    Err(_) => {}
                }
            }
            _ = tokio::time::sleep(sleep_duration) => {}
        }

        let position = state.interpolated_position();

        // Recheck queue 10s before track ends
        if config.merge_queue
            && state.is_playing
            && !state.queue_rechecked
            && state.current_duration > 10_000
            && position >= state.current_duration - 10_000
        {
            state.queue_rechecked = true;
            if let Ok(queue) = player.queue().await {
                let new_next_id = queue.first().map(|q| q.track_id.as_str());
                let old_next_id = if state.next_track_id.is_empty() {
                    None
                } else {
                    Some(state.next_track_id.as_str())
                };

                if new_next_id != old_next_id {
                    if let Some(start) = state.next_track_start {
                        state.display_lines.truncate(start.saturating_sub(1));
                    }
                    state.next_lyrics.clear();
                    state.next_track_id.clear();
                    state.next_track_start = None;

                    if let Some(next) = queue.into_iter().next() {
                        state.next_track_id = next.track_id;
                        state.next_artist = next.artist;
                        state.next_track_name = next.track;
                        if let Ok(nl) = provider
                            .fetch(&state.next_artist, &state.next_track_name)
                            .await
                        {
                            state.next_lyrics = nl;
                            ensure_leading_line(&mut state.next_lyrics);
                            state.next_track_start = append_next_track(
                                &mut state.display_lines,
                                &state.next_artist,
                                &state.next_track_name,
                                &state.next_lyrics,
                            );
                        }
                    }
                }
            }
        }

        // Auto-transition when position reaches track duration
        if state.is_playing
            && state.current_duration > 0
            && position >= state.current_duration
            && !state.next_track_id.is_empty()
        {
            let overflow = position - state.current_duration;

            if config.merge_queue && state.next_track_start.is_some() {
                state.current_track_offset = state.next_track_start.unwrap();
            } else {
                state.current_track_offset = 0;
                state.display_lines = SyncState::lyrics_to_display(&state.next_lyrics);
            }

            state.current_lyrics = std::mem::take(&mut state.next_lyrics);
            state.current_track_id.clone_from(&state.next_track_id);
            state.current_duration = 0;
            state.index = state.current_track_offset;
            state.error = None;
            state.queue_rechecked = false;

            state
                .fetch_next_track(&*player, &*provider, config.merge_queue)
                .await;

            state.last_state_position = overflow;
            state.last_state_time = Instant::now();
        }

        // Calculate current line index
        if !state.current_lyrics.is_empty() {
            let local_current = state.index.saturating_sub(state.current_track_offset);
            let clamped = local_current.min(state.current_lyrics.len().saturating_sub(1));
            let local_index = get_index(position, clamped, &state.current_lyrics);
            state.index = state.current_track_offset + local_index;
        }

        // Schedule next wake-up precisely at the next line transition
        sleep_duration = if state.is_playing && !state.current_lyrics.is_empty() {
            let local_idx = state.index.saturating_sub(state.current_track_offset);
            if local_idx + 1 < state.current_lyrics.len() {
                let next_time_ms = state.current_lyrics[local_idx + 1].time_ms;
                let pos_now = state.interpolated_position();
                if next_time_ms > pos_now {
                    Duration::from_millis(next_time_ms - pos_now).min(config.ui_timer_interval)
                } else {
                    Duration::from_millis(10)
                }
            } else {
                config.ui_timer_interval
            }
        } else {
            config.ui_timer_interval
        };

        if tx.send(state.to_update()).await.is_err() {
            break;
        }
    }
}

pub fn get_index(position_ms: u64, current_index: usize, lines: &[Line]) -> usize {
    if lines.len() <= 1 {
        return 0;
    }

    if position_ms >= lines[current_index].time_ms {
        for i in (current_index + 1)..lines.len() {
            if position_ms < lines[i].time_ms {
                return i - 1;
            }
        }
        return lines.len() - 1;
    }

    for i in (1..=current_index).rev() {
        if position_ms >= lines[i].time_ms {
            return i;
        }
    }
    0
}

pub fn append_next_track(
    display_lines: &mut Vec<DisplayLine>,
    artist: &str,
    track: &str,
    lyrics: &[Line],
) -> Option<usize> {
    if lyrics.is_empty() {
        return None;
    }
    display_lines.push(DisplayLine::Separator(format!("── {artist} - {track} ──")));
    let start = display_lines.len();
    display_lines.extend(lyrics.iter().cloned().map(DisplayLine::Lyric));
    Some(start)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::renderer::DisplayLine;

    fn make_lines(times: &[u64]) -> Vec<Line> {
        times
            .iter()
            .map(|&t| Line {
                time_ms: t,
                words: format!("line@{t}"),
            })
            .collect()
    }

    #[test]
    fn test_single_line() {
        let lines = make_lines(&[0]);
        assert_eq!(get_index(5000, 0, &lines), 0);
    }

    #[test]
    fn test_empty() {
        let lines = make_lines(&[]);
        assert_eq!(get_index(0, 0, &lines), 0);
    }

    #[test]
    fn test_before_first_line() {
        let lines = make_lines(&[1000, 2000, 3000]);
        assert_eq!(get_index(500, 0, &lines), 0);
    }

    #[test]
    fn test_exact_match_second_line() {
        let lines = make_lines(&[1000, 2000, 3000]);
        assert_eq!(get_index(2000, 0, &lines), 1);
    }

    #[test]
    fn test_between_lines() {
        let lines = make_lines(&[1000, 2000, 3000]);
        assert_eq!(get_index(2500, 0, &lines), 1);
    }

    #[test]
    fn test_after_last_line() {
        let lines = make_lines(&[1000, 2000, 3000]);
        assert_eq!(get_index(5000, 0, &lines), 2);
    }

    #[test]
    fn test_backward_search() {
        let lines = make_lines(&[1000, 2000, 3000]);
        assert_eq!(get_index(1500, 2, &lines), 0);
    }

    #[test]
    fn test_from_any_start_index() {
        let lines = make_lines(&[0, 1000, 2000, 3000, 4000]);
        for target in 0..lines.len() {
            let mid = if target + 1 < lines.len() {
                (lines[target].time_ms + lines[target + 1].time_ms) / 2
            } else {
                lines[target].time_ms + 500
            };
            for start in 0..lines.len() {
                assert_eq!(
                    get_index(mid, start, &lines),
                    target,
                    "get_index({mid}, {start}, ...) should be {target}"
                );
            }
        }
    }

    #[test]
    fn test_append_next_track_basic() {
        let mut lines: Vec<DisplayLine> = make_lines(&[1000, 2000])
            .into_iter()
            .map(DisplayLine::Lyric)
            .collect();
        let next = make_lines(&[0, 500]);
        let start = append_next_track(&mut lines, "Artist", "Song", &next);
        assert_eq!(start, Some(3));
        assert_eq!(lines.len(), 5);
        assert!(matches!(&lines[2], DisplayLine::Separator(s) if s.contains("Artist")));
        assert!(matches!(&lines[3], DisplayLine::Lyric(l) if l.time_ms == 0));
    }

    #[test]
    fn test_append_next_track_empty_lyrics() {
        let mut lines: Vec<DisplayLine> = make_lines(&[1000])
            .into_iter()
            .map(DisplayLine::Lyric)
            .collect();
        let result = append_next_track(&mut lines, "Artist", "Song", &[]);
        assert_eq!(result, None);
        assert_eq!(lines.len(), 1);
    }

    #[test]
    fn test_append_next_track_preserves_existing() {
        let mut lines: Vec<DisplayLine> = vec![
            DisplayLine::Lyric(Line {
                time_ms: 100,
                words: "old".into(),
            }),
            DisplayLine::Separator("── Old ──".into()),
            DisplayLine::Lyric(Line {
                time_ms: 200,
                words: "current".into(),
            }),
        ];
        let next = make_lines(&[0]);
        let start = append_next_track(&mut lines, "New", "Track", &next);
        assert_eq!(start, Some(4));
        assert_eq!(lines.len(), 5);
        assert_eq!(lines[0].text(), "old");
    }

    #[test]
    fn test_append_next_track_separator_format() {
        let mut lines: Vec<DisplayLine> = make_lines(&[1000])
            .into_iter()
            .map(DisplayLine::Lyric)
            .collect();
        let next = make_lines(&[0]);
        append_next_track(&mut lines, "Lord Huron", "The Night We Met", &next);
        if let DisplayLine::Separator(s) = &lines[1] {
            assert_eq!(s, "── Lord Huron - The Night We Met ──");
        } else {
            panic!("Expected Separator at index 1");
        }
    }
}

File: src/sync.ts
import { Channel } from "./channel";
import { ensureLeadingLine, type Line, type LyricsProvider } from "./lyrics";
import type { Player, State } from "./player";
import type { DisplayLine, Update } from "./renderer";

export interface SyncConfig {
  playerPollInterval: number;
  uiTimerInterval: number;
  mergeQueue: boolean;
}

export function defaultSyncConfig(): SyncConfig {
  return {
    playerPollInterval: 2000,
    uiTimerInterval: 200,
    mergeQueue: false,
  };
}

// Mirrors the Rust `Result<Option<State>>` value carried over the player-poll
// channel. `ok: false` is the `Err(_)` arm (which the loop ignores).
type PollResult = { ok: true; state: State | null } | { ok: false };

const sleep = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms));

class SyncState {
  currentTrackId = "";
  currentLyrics: Line[] = [];
  displayLines: DisplayLine[] = [];
  currentTrackOffset = 0;
  currentDuration = 0;
  nextLyrics: Line[] = [];
  nextArtist = "";
  nextTrackName = "";
  nextTrackId = "";
  nextTrackStart: number | null = null;
  index = 0;
  lastStatePosition = 0;
  lastStateTime: number = Date.now();
  isPlaying = false;
  error: string | null = null;
  queueRechecked = false;

  interpolatedPosition(): number {
    if (this.isPlaying) {
      return this.lastStatePosition + (Date.now() - this.lastStateTime);
    }
    return this.lastStatePosition;
  }

  static lyricsToDisplay(lyrics: Line[]): DisplayLine[] {
    return lyrics.map((line) => ({ kind: "lyric", line }));
  }

  clear(): void {
    this.currentTrackId = "";
    this.currentLyrics = [];
    this.displayLines = [];
    this.nextLyrics = [];
    this.nextTrackId = "";
    this.nextTrackStart = null;
    this.currentTrackOffset = 0;
    this.currentDuration = 0;
    this.index = 0;
    this.error = null;
  }

  toUpdate(): Update {
    return {
      lines: this.displayLines.slice(),
      index: this.index,
      error: this.error,
    };
  }

  async fetchNextTrack(
    player: Player,
    provider: LyricsProvider,
    mergeQueue: boolean,
  ): Promise<void> {
    this.nextLyrics = [];
    this.nextTrackId = "";
    this.nextTrackStart = null;
    let queue;
    try {
      queue = await player.queue();
    } catch {
      return;
    }
    const next = queue[0];
    if (next === undefined) {
      return;
    }
    this.nextTrackId = next.trackId;
    this.nextArtist = next.artist;
    this.nextTrackName = next.track;
    let nl: Line[];
    try {
      nl = await provider.fetch(this.nextArtist, this.nextTrackName);
    } catch {
      return;
    }
    this.nextLyrics = nl;
    ensureLeadingLine(this.nextLyrics);
    if (mergeQueue) {
      this.nextTrackStart = appendNextTrack(
        this.displayLines,
        this.nextArtist,
        this.nextTrackName,
        this.nextLyrics,
      );
    }
  }

  async transitionToNewTrack(
    trackId: string,
    artist: string,
    track: string,
    provider: LyricsProvider,
    player: Player,
    smooth: boolean,
    mergeQueue: boolean,
  ): Promise<void> {
    if (smooth) {
      // `unwrap()`: smooth is only true when nextTrackStart is set.
      this.currentTrackOffset = this.nextTrackStart as number;
      this.currentLyrics = this.nextLyrics;
      this.nextLyrics = [];
    } else if (trackId === this.nextTrackId && this.nextLyrics.length > 0) {
      this.currentTrackOffset = 0;
      this.currentLyrics = this.nextLyrics;
      this.nextLyrics = [];
      this.displayLines = SyncState.lyricsToDisplay(this.currentLyrics);
    } else {
      this.currentTrackOffset = 0;
      let l: Line[];
      try {
        l = await provider.fetch(artist, track);
      } catch (e) {
        this.currentLyrics = [];
        this.displayLines = [];
        this.nextLyrics = [];
        this.nextTrackId = "";
        this.nextTrackStart = null;
        this.error = e instanceof Error ? e.message : String(e);
        this.currentTrackId = trackId;
        this.index = 0;
        return;
      }
      this.currentLyrics = l;
      ensureLeadingLine(this.currentLyrics);
      this.displayLines = SyncState.lyricsToDisplay(this.currentLyrics);
    }

    this.currentTrackId = trackId;
    this.index = this.currentTrackOffset;
    this.error = null;

    await this.fetchNextTrack(player, provider, mergeQueue);
  }
}

export function startSync(
  player: Player,
  provider: LyricsProvider,
  config: SyncConfig,
): Channel<Update> {
  const tx = new Channel<Update>();
  // Spawn the loop; do NOT await.
  void syncLoop(player, provider, config, tx);
  return tx;
}

async function syncLoop(
  player: Player,
  provider: LyricsProvider,
  config: SyncConfig,
  tx: Channel<Update>,
): Promise<void> {
  const playerChan = new Channel<PollResult>();

  // Player-poll task: poll `state()`, push the result, sleep, repeat. Breaks
  // when the channel is closed (mirrors `send(...).is_err()`).
  void (async () => {
    while (true) {
      let result: PollResult;
      try {
        const state = await player.state();
        result = { ok: true, state };
      } catch {
        result = { ok: false };
      }
      if (!playerChan.send(result)) {
        break;
      }
      await sleep(config.playerPollInterval);
    }
  })();

  const state = new SyncState();
  let sleepDuration = config.uiTimerInterval;

  // Stable-promise race: hold `recvP` across iterations and renew it ONLY when
  // the player branch resolves.
  let recvP = playerChan.recv();

  while (true) {
    const winner = await Promise.race([
      recvP.then((v) => ({ tag: "msg", v }) as const),
      sleep(sleepDuration).then(() => ({ tag: "tick" }) as const),
    ]);

    if (winner.tag === "msg") {
      const result = winner.v;
      recvP = playerChan.recv();
      // `Some(result) = player_rx.recv()`: `undefined` means closed → skip.
      if (result !== undefined) {
        if (result.ok && result.state !== null) {
          const playerState = result.state;
          state.isPlaying = playerState.isPlaying;
          state.currentDuration = playerState.durationMs;

          if (playerState.trackId !== state.currentTrackId) {
            const smooth =
              playerState.trackId === state.nextTrackId &&
              state.nextTrackStart !== null;

            await state.transitionToNewTrack(
              playerState.trackId,
              playerState.artist,
              playerState.track,
              provider,
              player,
              smooth && config.mergeQueue,
              config.mergeQueue,
            );
            state.queueRechecked = false;
          }

          state.lastStatePosition = playerState.positionMs;
          state.lastStateTime = Date.now();
        } else if (result.ok && result.state === null) {
          state.isPlaying = false;
          if (state.currentTrackId !== "") {
            state.clear();
          }
        }
        // result.ok === false → Err(_) arm: do nothing.
      }
    }
    // tick branch: just loop; do NOT recreate recvP.

    const position = state.interpolatedPosition();

    // Recheck queue 10s before track ends.
    if (
      config.mergeQueue &&
      state.isPlaying &&
      !state.queueRechecked &&
      state.currentDuration > 10000 &&
      position >= state.currentDuration - 10000
    ) {
      state.queueRechecked = true;
      let queue;
      try {
        queue = await player.queue();
      } catch {
        queue = null;
      }
      if (queue !== null) {
        const firstItem = queue[0];
        const newNextId: string | null =
          firstItem !== undefined ? firstItem.trackId : null;
        const oldNextId: string | null =
          state.nextTrackId === "" ? null : state.nextTrackId;

        if (newNextId !== oldNextId) {
          if (state.nextTrackStart !== null) {
            const truncLen = Math.max(state.nextTrackStart - 1, 0);
            state.displayLines.length = Math.min(
              truncLen,
              state.displayLines.length,
            );
          }
          state.nextLyrics = [];
          state.nextTrackId = "";
          state.nextTrackStart = null;

          const next = queue[0];
          if (next !== undefined) {
            state.nextTrackId = next.trackId;
            state.nextArtist = next.artist;
            state.nextTrackName = next.track;
            let nl: Line[] | null;
            try {
              nl = await provider.fetch(state.nextArtist, state.nextTrackName);
            } catch {
              nl = null;
            }
            if (nl !== null) {
              state.nextLyrics = nl;
              ensureLeadingLine(state.nextLyrics);
              state.nextTrackStart = appendNextTrack(
                state.displayLines,
                state.nextArtist,
                state.nextTrackName,
                state.nextLyrics,
              );
            }
          }
        }
      }
    }

    // Auto-transition when position reaches track duration.
    if (
      state.isPlaying &&
      state.currentDuration > 0 &&
      position >= state.currentDuration &&
      state.nextTrackId !== ""
    ) {
      const overflow = position - state.currentDuration;

      if (config.mergeQueue && state.nextTrackStart !== null) {
        state.currentTrackOffset = state.nextTrackStart;
      } else {
        state.currentTrackOffset = 0;
        state.displayLines = SyncState.lyricsToDisplay(state.nextLyrics);
      }

      state.currentLyrics = state.nextLyrics;
      state.nextLyrics = [];
      state.currentTrackId = state.nextTrackId;
      state.currentDuration = 0;
      state.index = state.currentTrackOffset;
      state.error = null;
      state.queueRechecked = false;

      await state.fetchNextTrack(player, provider, config.mergeQueue);

      state.lastStatePosition = overflow;
      state.lastStateTime = Date.now();
    }

    // Calculate current line index.
    if (state.currentLyrics.length > 0) {
      const localCurrent = Math.max(state.index - state.currentTrackOffset, 0);
      const clamped = Math.min(
        localCurrent,
        Math.max(state.currentLyrics.length - 1, 0),
      );
      const localIndex = getIndex(position, clamped, state.currentLyrics);
      state.index = state.currentTrackOffset + localIndex;
    }

    // Schedule next wake-up precisely at the next line transition.
    if (state.isPlaying && state.currentLyrics.length > 0) {
      const localIdx = Math.max(state.index - state.currentTrackOffset, 0);
      if (localIdx + 1 < state.currentLyrics.length) {
        const nextTimeMs = state.currentLyrics[localIdx + 1]!.timeMs;
        const posNow = state.interpolatedPosition();
        if (nextTimeMs > posNow) {
          sleepDuration = Math.min(nextTimeMs - posNow, config.uiTimerInterval);
        } else {
          sleepDuration = 10;
        }
      } else {
        sleepDuration = config.uiTimerInterval;
      }
    } else {
      sleepDuration = config.uiTimerInterval;
    }

    if (!tx.send(state.toUpdate())) {
      break;
    }
  }

  playerChan.close();
}

export function getIndex(
  positionMs: number,
  currentIndex: number,
  lines: Line[],
): number {
  if (lines.length <= 1) {
    return 0;
  }

  if (positionMs >= lines[currentIndex]!.timeMs) {
    for (let i = currentIndex + 1; i < lines.length; i++) {
      if (positionMs < lines[i]!.timeMs) {
        return i - 1;
      }
    }
    return lines.length - 1;
  }

  for (let i = currentIndex; i >= 1; i--) {
    if (positionMs >= lines[i]!.timeMs) {
      return i;
    }
  }
  return 0;
}

export function appendNextTrack(
  displayLines: DisplayLine[],
  artist: string,
  track: string,
  lyrics: Line[],
): number | null {
  if (lyrics.length === 0) {
    return null;
  }
  displayLines.push({ kind: "separator", text: `── ${artist} - ${track} ──` });
  const start = displayLines.length;
  for (const line of lyrics) {
    displayLines.push({ kind: "lyric", line });
  }
  return start;
}
